fava/api/__init__.py

```diff
@@ -424,3 +433,29 @@ def account_metadata(self, account_name):
             if isinstance(posting, Open):
                 return posting.meta
         return {}
+
+    def entries_for_link(self, link):
+        """Entries with the specified link."""
+        return [entry for entry in self.entries
+                if isinstance(entry, (Transaction, Document)) and
+                entry.links and link in entry.links]
+
+    def insert_metadata(self, filepath, lineno, basekey, value):
+        """Insert metadata into a file at lineno. Also, prevent duplicate
+        keys."""
+
+        key = None
+        for entry in self.entries:
+            if isinstance(entry, Transaction):
+                if entry.meta['filename'] == filepath and \
+                   entry.meta['lineno'] == lineno:
+                    key = next_key(basekey, entry.meta)
+                    break
+                for pos in entry.postings:
+                    if pos.meta and pos.meta['filename'] == filepath and \
+                       pos.meta['lineno'] == lineno:
+                        key = next_key(basekey, pos.meta if pos.meta else [])
+                        break
+
+        assert key != None
```

Comment:
Actually this should raise an Exception that is properly handled in the application